### PR TITLE
[GrandCentralTaps] Move (new) MemTaps to WiringProblem.

### PR DESCRIPTION
**Pull request by dtzSiFive**
While visiting add small check before dereferencing an Optional.

### CODE REVIEW

File: lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp [hunk 1]
```diff
@@ -778,45 +778,22 @@ LogicalResult applyGCTMemTapsWithWires(const AnnoPathValue &target,
...
-      dyn_cast<FModuleOp>(srcTarget->ref.getModule().getOperation());
-  ImplicitLocOpBuilder refSendBuilder(srcModule.getLoc(), srcModule);
+  if (!wireTarget)
+    return mlir::emitError(loc, "Annotation '" + Twine(memTapClass) +
+                                    "' with path '.taps[0]' contains target '" +
+                                    wireTargetStr +
+                                    "' that cannot be resolved.")
+               .attachNote()
+           << "The full Annotation is reproduced here: " << anno << "\n";
+
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp:788]: 👍 Thanks for adding this. I had to check that `.taps[0]` made sense to include and I think it does. Nice.

File: lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp [hunk 2]
```diff
@@ -778,45 +778,22 @@ LogicalResult applyGCTMemTapsWithWires(const AnnoPathValue &target,
...
-  if (wireTarget->ref.getOp()->getResult(0).getType() != refResolve.getType())
+  if (wireTarget->ref.getOp()->getResult(0).getType() !=
+      cast<RefType>(sendVal.getType()).getType())
     return wireTarget->ref.getOp()->emitError(
         "cannot generate the MemTap, wiretap Type does not match the memory "
         "type");
-  refResolveBuilder.create<StrictConnectOp>(
-      wireTarget->ref.getOp()->getResult(0), refResolve.getResult());
+  auto sink = wireTarget->ref.getOp()->getResult(0);
+  state.wiringProblems.push_back({sendVal, sink, "memTap"});
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp:796]: This two-line code is the `WiringProblem` dream. I'm glad it looks like we have a clean API for these types of problems.

**Review by seldridge — APPROVED** [review]

File: lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp (hunk 1 above)
Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp:788]: I think a helper for emitting these errors would be very useful.   Avoid deviations/typos and make it less verbose to emit these errors on failure paths.

I was doubtful about the "path" portion of some of these errors earlier to be honest (confused why we appended '.class' to one, for example) but yeah I think they're useful now.  If nothing else consistency here is good :+1: .

**Review by prithayan — APPROVED** [review]
Thanks for fixing this, Schuyler's wiringProblem seems awesome :)

**Pull request merged**